idea/idea-frontend-api/src/org/jetbrains/kotlin/idea/frontend/api/components/KtMemberSymbolProvider.kt

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.idea.frontend.api.components
+
+import org.jetbrains.kotlin.descriptors.Modality
+import org.jetbrains.kotlin.idea.frontend.api.symbols.KtCallableSymbol
+import org.jetbrains.kotlin.idea.frontend.api.symbols.KtClassOrObjectSymbol
+import org.jetbrains.kotlin.idea.frontend.api.symbols.markers.KtSymbolWithModality
+
+abstract class KtMemberSymbolProvider : KtAnalysisSessionComponent() {
+    /**
+     * Processes member symbols.
+     * @param collect callback function that is invoked with member symbol and related information:
+     *   - symbol: the member symbol
+     *   - isNotImplemented: the member symbol is missing an implementation
+     *   - isAmbiguouslyImplemented: the member symbol is an intersection of multiple non-abstract super symbols
+     *   - isOverridable: the member symbol is open and not implemented in the current class and hence can be implemented
+     *   - intersectionSymbols: if the member symbol is an intersection of multiple symbols from multiple super classes, then this contains
+     *     all the symbols in the intersection. It's set to `null` if `symbol` is not an intersection.
+     */
+    abstract fun processMemberSymbols(
```

Comment:
We already have a similar concept which is called `KtMemberScope`, can we somehow reuse it?

Reply:
I struggled about where to put this. So would you prefer to have a data class `MemberSymbolWithOverrideInfo` and add a method to return a sequence of that to `KtMemberScope`?

Reply:
Actually I just changed how this is implemented. Now I am using KtMemberScope to enumerate and added a `KtOverrideInfoProvider` that provides the necessary metadata.